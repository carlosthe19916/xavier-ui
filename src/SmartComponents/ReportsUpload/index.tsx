import React, { createRef } from 'react';
import { bindActionCreators } from 'redux';
import { withRouter, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import {
    Button,
    Modal,
    Form,
    FormGroup,
    TextInput,
    ActionGroup,
    InputGroup,
    InputGroupText,
    Grid,
    GridItem,
    ButtonVariant,
    Split,
    SplitItem,
    Stack,
    StackItem,
    Bullseye,
    EmptyState,
    EmptyStateVariant,
    EmptyStateIcon,
    Title,
    EmptyStateBody,
    ProgressMeasureLocation,
    Progress,
    TitleLevel,
    ProgressVariant,
    EmptyStateSecondaryActions
} from '@patternfly/react-core';
import { EditIcon, CubesIcon } from '@patternfly/react-icons';
import * as uploadActions from '../../actions/UploadActions';
import { GlobalState } from '../../models/state';
import { RouterGlobalProps } from '../../models/router';
import {
    Main,
    PageHeader,
    PageHeaderTitle
} from '@redhat-cloud-services/frontend-components';
import { Upload } from '../../models';
import Dropzone from 'react-dropzone';

interface StateToProps {
    file: File | null;
    success: boolean | null;
    error: string | null;
    progress: number;
    uploading: boolean
}

interface DispatchToProps {
    uploadProgress: (progress: number) => void;
    uploadRequest: (upload: Upload, config: {}) => void;
}

interface Props extends StateToProps, DispatchToProps, RouterGlobalProps {
};

interface State {
    submitted: boolean,

    file?: File | null,
    reportName?: string,
    reportDescription?: string | null,
    yearOverYearGrowthRatePercentage?: number | null,
    percentageOfHypervisorsMigratedOnYear1?: number | null,
    percentageOfHypervisorsMigratedOnYear2?: number | null,
    percentageOfHypervisorsMigratedOnYear3?: number | null
};

export class ReportsUpload extends React.Component<Props, State> {

    selectFileButtonRef: any;

    constructor(props: Props) {
        super(props);
        this.state = {
            submitted: false,

            file: this.props.file,
            reportName: this.props.file ? this.props.file.name : '',
            reportDescription: '',
            yearOverYearGrowthRatePercentage: 5,
            percentageOfHypervisorsMigratedOnYear1: 50,
            percentageOfHypervisorsMigratedOnYear2: 30,
            percentageOfHypervisorsMigratedOnYear3: 10
        };
        this.handleModalToggle = this.handleModalToggle.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.onDrop = this.onDrop.bind(this);
    }

    handleModalToggle () {
        this.props.history.push('/reports');
    };

    handleInputChange(value: string, event: any) {
        const target = event.target;
        const name = target.name;

        this.setState({
            [name]: value
        });
    }

    handleSubmit(event: any) {
        event.preventDefault();
        this.setState({
            submitted: true
        });

        const upload: Upload = {
            ... this.state
        };

        const config = {
            headers: {
                'X-Requested-With': 'XMLHttpRequest'
            },
            onUploadProgress: (progressEvent: any) => {
                const progress: number = Math.round((progressEvent.loaded / progressEvent.total) * 100);
                this.props.uploadProgress(progress);
            }
        };

        this.props.uploadRequest(upload, config);
    }

    progress() {
        return (
            <Bullseye>
                <EmptyState variant={ EmptyStateVariant.full }>
                    <EmptyStateIcon icon={ CubesIcon } />
                    <Title headingLevel={ TitleLevel.h5 } size="lg">
                        Empty State
                    </Title>
                    <div className="pf-c-empty-state__body">
                        <Progress
                            value={ this.props.progress }
                            measureLocation={ ProgressMeasureLocation.outside }
                            variant={ this.props.error ? ProgressVariant.danger : ProgressVariant.info }
                        />
                    </div>
                    <EmptyStateBody>
                        Your file is been uploaded, the process can take some time.
                    </EmptyStateBody>
                    <EmptyStateSecondaryActions>
                        {
                            this.props.success && <Button
                                variant={ ButtonVariant.primary }
                                component={ Link }
                                to={ '/reports' }>
                                Close
                            </Button>
                        }
                    </EmptyStateSecondaryActions>
                </EmptyState>
            </Bullseye>
        );
    }

    onDrop(files: File[]): void {
        this.setState({
            file: files[0],
            reportName: files[0].name
        });
    };

    form() {
        const dropzoneRef: any = createRef();
        const openDialog = () => {
            if (dropzoneRef.current) {
                dropzoneRef.current.open();
            }
        };

        return (
            <Form onSubmit={ this.handleSubmit }>
                <FormGroup
                    fieldId="file"
                    label=""
                >
                    <Dropzone
                        onDrop={ this.onDrop }
                        ref={ dropzoneRef }
                        noClick noKeyboard
                        multiple={ false }
                        accept={ [ 'application/zip', 'application/json' ] }
                    >
                        { ({ getRootProps, getInputProps }) => {
                            return (
                                <div className="container">
                                    <div { ...getRootProps({ className: 'dropzone' }) }>
                                        <input { ...getInputProps() } />
                                        <InputGroup>
                                            <TextInput
                                                id="file"
                                                name="file"
                                                type="text"
                                                aria-label="Select a File"
                                                value={ this.state.file ? this.state.file.name : '' } />
                                            <Button
                                                variant={ ButtonVariant.tertiary }
                                                aria-label="search button for search input"
                                                onClick={ openDialog }
                                            >
                                                <EditIcon />
                                            </Button>
                                        </InputGroup>
                                    </div>
                                </div>
                            );
                        } }
                    </Dropzone>
                </FormGroup>
                <FormGroup
                    isRequired
                    label="Report name"
                    fieldId="reportName"
                >
                    <TextInput
                        isRequired
                        type="text"
                        id="reportName"
                        name="reportName"
                        aria-describedby="Report name"
                        value={ this.state.reportName }
                        onChange={ this.handleInputChange }
                    />
                </FormGroup>
                <FormGroup
                    label="Report description"
                    fieldId="reportDescription"
                >
                    <TextInput
                        isRequired
                        type="text"
                        id="reportDescription"
                        name="reportDescription"
                        aria-describedby="Report description"
                        value={ this.state.reportDescription }
                        onChange={ this.handleInputChange }
                    />
                </FormGroup>
                <FormGroup
                    isRequired
                    label="Year-over-year growth rate for new hypervisors"
                    fieldId="yearOverYearGrowthRatePercentage"
                >
                    <Grid>
                        <GridItem span={ 2 }>
                            <InputGroup>
                                <TextInput
                                    isRequired
                                    type="number"
                                    id="yearOverYearGrowthRatePercentage"
                                    name="yearOverYearGrowthRatePercentage"
                                    value={ this.state.yearOverYearGrowthRatePercentage }
                                    onChange={ this.handleInputChange }
                                />
                                <InputGroupText>%</InputGroupText>
                            </InputGroup>
                        </GridItem>
                    </Grid>
                </FormGroup>
                <FormGroup
                    label="Percentage of hypervisors migrated each year"
                    fieldId=""
                >
                    <Split>
                        <SplitItem isFilled={ false }>
                            <Stack gutter="sm">
                                <StackItem isFilled={ false }>
                                    <InputGroup>
                                        <InputGroupText>Year 1</InputGroupText>
                                        <TextInput
                                            isRequired
                                            type="number"
                                            id="percentageOfHypervisorsMigratedOnYear1"
                                            name="percentageOfHypervisorsMigratedOnYear1"
                                            aria-label="Percentage of hypervisors migrated on year 1"
                                            value={ this.state.percentageOfHypervisorsMigratedOnYear1 }
                                            onChange={ this.handleInputChange }
                                        />
                                        <InputGroupText>%</InputGroupText>
                                    </InputGroup>
                                </StackItem>
                                <StackItem isFilled={ false }>
                                    <InputGroup>
                                        <InputGroupText>Year 2</InputGroupText>
                                        <TextInput
                                            isRequired
                                            type="number"
                                            id="percentageOfHypervisorsMigratedOnYear2"
                                            name="percentageOfHypervisorsMigratedOnYear2"
                                            aria-label="Percentage of hypervisors migrated on year 2"
                                            value={ this.state.percentageOfHypervisorsMigratedOnYear2 }
                                            onChange={ this.handleInputChange }
                                        />
                                        <InputGroupText>%</InputGroupText>
                                    </InputGroup>
                                </StackItem>
                                <StackItem isFilled={ false }>
                                    <InputGroup>
                                        <InputGroupText>Year 3</InputGroupText>
                                        <TextInput
                                            isRequired
                                            type="number"
                                            id="percentageOfHypervisorsMigratedOnYear3"
                                            name="percentageOfHypervisorsMigratedOnYear3"
                                            aria-label="Percentage of hypervisors migrated on year 3"
                                            value={ this.state.percentageOfHypervisorsMigratedOnYear3 }
                                            onChange={ this.handleInputChange }
                                        />
                                        <InputGroupText>%</InputGroupText>
                                    </InputGroup>
                                </StackItem>
                            </Stack>
                        </SplitItem>
                    </Split>
                </FormGroup>
                <ActionGroup>
                    <Button variant="primary" type="submit">Create report</Button>
                    <Button variant="secondary" component={ Link } to={ '/reports' }>Cancel</Button>
                </ActionGroup>
            </Form>
        );
    }

    render() {
        return (
            <React.Fragment>
                <PageHeader>
                    <PageHeaderTitle title='Reports' />
                </PageHeader>
                <Main>
                    <Modal
                        title='Report options'
                        isLarge
                        isOpen={ true }
                        onClose={ this.handleModalToggle }
                        ariaDescribedById="no-header-example"
                        actions={ [] }
                    >
                        { this.state.submitted ? this.progress() : this.form() }
                    </Modal>
                </Main>
            </React.Fragment>
        );
    }
}

const mapStateToProps = (state: GlobalState)  => {
    let {
        uploadState: {
            file,
            success,
            error,
            progress,
            uploading
        }
    } = state;
    return {
        file,
        success,
        error,
        progress,
        uploading
    };
};

const mapDispatchToProps = (dispatch: any) =>
    bindActionCreators({
        uploadRequest: uploadActions.uploadRequest,
        uploadProgress: uploadActions.uploadProgress
    }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(ReportsUpload));
