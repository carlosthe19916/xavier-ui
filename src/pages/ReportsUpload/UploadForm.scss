.percentage-box {
    border-top: 0 !important;
    border-right: 0 !important;
    border-bottom: 0 !important;
    color: inherit !important;
}

.year-box {
    border-top: 0 !important;
    border-left: 0 !important;
    border-bottom: 0 !important;
    color: inherit !important;
}
