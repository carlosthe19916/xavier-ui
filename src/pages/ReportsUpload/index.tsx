import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import ReportsUpload from './ReportsUpload';
import * as uploadActions from '../../actions/UploadActions';
import { GlobalState } from '../../models/state';

const mapStateToProps = (state: GlobalState)  => {
    const {
        uploadState: {
            file,
            success,
            error,
            progress,
            uploading
        }
    } = state;
    return {
        file,
        success,
        error,
        progress,
        uploading
    };
};

const mapDispatchToProps = (dispatch: any) =>
    bindActionCreators({
        uploadRequest: uploadActions.uploadRequest,
        uploadProgress: uploadActions.uploadProgress,
        selectUploadFile: uploadActions.selectUploadFile
    }, dispatch);

export default withRouter(
    connect(
        mapStateToProps,
        mapDispatchToProps
    )(ReportsUpload)
);
