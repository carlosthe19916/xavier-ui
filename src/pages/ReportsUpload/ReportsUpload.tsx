import React from 'react';
import {
    Button,
    Modal,
    ButtonVariant,
    Bullseye,
    EmptyState,
    EmptyStateVariant,
    EmptyStateIcon,
    Title,
    EmptyStateBody,
    ProgressMeasureLocation,
    Progress,
    TitleLevel,
    ProgressVariant,
    EmptyStateSecondaryActions
} from '@patternfly/react-core';
import { VolumeIcon } from '@patternfly/react-icons';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { RouterGlobalProps } from '../../models/router';
import {
    Main,
    PageHeader,
    PageHeaderTitle
} from '@redhat-cloud-services/frontend-components';
import { Upload } from '../../models';
import './ReportsUpload.scss';
import UploadForm from './UploadForm';
import { validateForm } from '../../Utilities/formUtils';

interface FormValues {
    file: string;
    reportName: string;
    reportDescription: string;
    yearOverYearGrowthRatePercentage: number;
    percentageOfHypervisorsMigratedOnYear1: number;
    percentageOfHypervisorsMigratedOnYear2: number;
    percentageOfHypervisorsMigratedOnYear3: number;
    percentageOfHypervisorsMigratedSum: number // Sum of year1, year2, and year3
}

interface StateToProps {
    file: File;
    success: boolean | null;
    error: string | null;
    progress: number;
    uploading: boolean
}

interface DispatchToProps {
    uploadProgress: (progress: number) => void;
    uploadRequest: (upload: Upload, config: {}) => void;
    selectUploadFile: (file: File) => void;
}

interface Props extends StateToProps, DispatchToProps, RouterGlobalProps {
}

interface State {
    showForm: boolean;
    timeoutToRedirect: number;
}

const initialFormValue: FormValues = {
    file: '',
    reportName: '',
    reportDescription: '',
    yearOverYearGrowthRatePercentage: 5,
    percentageOfHypervisorsMigratedOnYear1: 50,
    percentageOfHypervisorsMigratedOnYear2: 30,
    percentageOfHypervisorsMigratedOnYear3: 10,
    percentageOfHypervisorsMigratedSum: 90
};

const formValidationSchema = (values: FormValues) => {
    return Yup.object().shape({
        file: Yup.string()
        .required('File is mandatory'),
        reportName: Yup.string()
        .min(3, 'Report name must contain at least 3 characters ')//reproduce
        .max(250, 'Report name must contain fewer than 250 characters')
        .required('Report name is mandatory'),
        reportDescription: Yup.string()
        .max(250, 'Report description must contain fewer than 250 characters'),
        yearOverYearGrowthRatePercentage: Yup.number()
        .typeError('Invalid number')
        .min(0, 'Value must be greater than or equal to 0')
        .required('Growth rate percentage is mandatory'),
        percentageOfHypervisorsMigratedOnYear1: Yup.number()
        .typeError('Invalid number')
        .min(0, 'Value must be greater than or equal to 0')
        .max(100, 'Value must be less than or equal to 100')
        .required('Percentage of hypervisors migrated is mandatory'),
        percentageOfHypervisorsMigratedOnYear2: Yup.number()
        .typeError('Invalid number')
        .min(0, 'Vaaslue must be greater than or equal to 0')
        .max(100, 'Value must be less than or equal to 100')
        .required('Percentage of hypervisors migrated is mandatory'),
        percentageOfHypervisorsMigratedOnYear3: Yup.number()
        .typeError('Invalid number')
        .min(0, 'Value must be greater than or equal to 0')
        .max(100, 'Value must be less than or equal to 100')
        .required('Percentage of hypervisors migrated is mandatory'),
        percentageOfHypervisorsMigratedSum: Yup.number()
        .typeError('Invalid number')
        .min(0, 'Value must be greater than or equal to 0')
        .max(100, 'Value must be less than or equal to 100')
        .test('validSum', 'The total percentage must not exceed 100', () => {
            const sum = values.percentageOfHypervisorsMigratedOnYear1 +
                values.percentageOfHypervisorsMigratedOnYear2 +
                values.percentageOfHypervisorsMigratedOnYear3;
            return sum >= 0 && sum <= 100;
        })
    });
};

class ReportsUpload extends React.Component<Props, State> {

    redirectTimer: any;
    beforeUnloadHandler: any;

    constructor(props: Props) {
        super(props);

        this.state = {
            showForm: true,
            timeoutToRedirect: 3
        };

        this.redirectTimer = null;
    }

    componentDidMount() {
        this.beforeUnloadHandler = window.addEventListener('beforeunload', (event) => {
            if (this.props.uploading) {
                event.preventDefault();
                return event.returnValue = 'Are you sure you want to close? The upload has not finished yet';
            }

            return event;
        });
    }

    componentDidUpdate(_prevProps: Props, prevState: State) {
        if (this.props.success && prevState.timeoutToRedirect !== 0 && this.state.timeoutToRedirect === 0) {
            this.props.history.push('/reports');
        }
    }

    componentWillUnmount() {
        window.removeEventListener('beforeunload', this.beforeUnloadHandler);
    }

    /**
     * Called by "Close modal" or by the "Cancel button"
     */
    handleCloseModel = () => {
        if (!this.props.uploading) {
            this.props.history.push('/reports');
        }
    };

    actionsOnUploadSuccess = () => {
        const { timeoutToRedirect } = this.state;

        this.startRedirectTimer();

        return (
            <Button type="button" variant={ ButtonVariant.primary } onClick={ this.handleCloseModel }> Closing in { timeoutToRedirect } </Button>
        );
    }

    startRedirectTimer = () => {
        if (this.redirectTimer === null && this.state.timeoutToRedirect > 0) {
            this.redirectTimer = setInterval(this.redirectCountDown, 1000);
        }
    }

    redirectCountDown = () => {
        let timeoutToRedirect = this.state.timeoutToRedirect - 1;
        this.setState({
            timeoutToRedirect
        });

        if (timeoutToRedirect === 0) {
            clearInterval(this.redirectTimer);
        }
    }

    handleFormSubmit = (values: FormValues) => {
        this.setState({
            showForm: false
        });

        const upload: Upload = {
            file: this.props.file,
            reportName: values.reportName,
            reportDescription: values.reportDescription,
            yearOverYearGrowthRatePercentage: values.yearOverYearGrowthRatePercentage,
            percentageOfHypervisorsMigratedOnYear1: values.percentageOfHypervisorsMigratedOnYear1,
            percentageOfHypervisorsMigratedOnYear2: values.percentageOfHypervisorsMigratedOnYear2,
            percentageOfHypervisorsMigratedOnYear3: values.percentageOfHypervisorsMigratedOnYear3
        };

        const config = {
            headers: {
                'X-Requested-With': 'XMLHttpRequest'
            },
            onUploadProgress: (progressEvent: any) => {
                const progress: number = Math.round((progressEvent.loaded / progressEvent.total) * 100);
                this.props.uploadProgress(progress);
            }
        };

        this.props.uploadRequest(upload, config);
    }

    onFileSelected = (files: File[]): void => {
        this.props.selectUploadFile(files[0]);
    };

    renderProgress() {
        return (
            <Bullseye>
                <EmptyState variant={ EmptyStateVariant.full }>
                    <EmptyStateIcon icon={ VolumeIcon } />
                    <Title headingLevel={ TitleLevel.h5 } size="lg">
                        Upload
                    </Title>
                    <div className="pf-c-empty-state__body">
                        <Progress
                            value={ this.props.progress }
                            measureLocation={ ProgressMeasureLocation.outside }
                            variant={ this.props.error ? ProgressVariant.danger : ProgressVariant.info }
                        />
                    </div>
                    <EmptyStateBody>
                        {
                            this.props.success ?
                                'Finished successfully. We will redirect you to the next page.' :
                                'Your file is been uploaded, the process can take some time.'
                        }
                    </EmptyStateBody>
                    <EmptyStateSecondaryActions>
                        {
                            this.props.success ? this.actionsOnUploadSuccess() : ''
                        }
                    </EmptyStateSecondaryActions>
                </EmptyState>
            </Bullseye>
        );
    }

    renderForm() {
        return (
            <Formik
                initialValues={ initialFormValue }
                validate={ (values) => validateForm(values, formValidationSchema) }
                onSubmit={ this.handleFormSubmit }
            >
                {
                    props => <UploadForm
                        onFileSelected={ this.onFileSelected }
                        file={ this.props.file }
                        handleCancel={ this.handleCloseModel }
                        { ...props } />
                }
            </Formik>
        );
    }

    render() {
        return (
            <React.Fragment>
                <PageHeader>
                    <PageHeaderTitle title='Reports' />
                </PageHeader>
                <Main>
                    <Modal
                        title='Report options'
                        isLarge
                        isOpen={ true }
                        onClose={ this.handleCloseModel }
                        ariaDescribedById="Report options"
                        actions={ [] }
                    >
                        { this.state.showForm ? this.renderForm() : this.renderProgress() }
                    </Modal>
                </Main>
            </React.Fragment>
        );
    }
}

export default ReportsUpload;
