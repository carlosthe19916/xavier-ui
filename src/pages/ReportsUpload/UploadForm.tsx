import React from 'react';
import {
    FormikState,
    FormikValues,
    FormikActions,
    FormikHandlers
} from 'formik';
import {
    Form,
    FormGroup,
    InputGroup,
    TextInput,
    Button,
    ButtonVariant,
    Gallery,
    GalleryItem,
    InputGroupText,
    Stack,
    StackItem,
    ActionGroup
} from '@patternfly/react-core';
import Dropzone from 'react-dropzone';
import './UploadForm.scss';

interface UploadFormProps extends FormikState<FormikValues>, FormikActions<FormikValues>, FormikHandlers {
    file: File;
    onFileSelected: Function;
    handleCancel: (e: any) => any;
}

class UploadForm extends React.Component<UploadFormProps, { }> {

    constructor(props: UploadFormProps) {
        super(props);
    }

    componentDidMount() {
        this.updateFieldValues();
    }

    componentDidUpdate(prevProps: any) {
        if (prevProps.file !== this.props.file) {
            this.updateFieldValues();
        }
    }

    updateFieldValues() {
        if (this.props.file) {
            const fileName = this.props.file.name;
            setTimeout(() => {
                this.props.setFieldValue('file', fileName);
                if (!this.props.values.reportName) {
                    this.props.setFieldValue('reportName', fileName);
                }
            }, 0);
        }
    }

    render() {
        const {
            values,
            errors,
            touched,
            handleChange,
            handleBlur,
            handleSubmit,
            isSubmitting
        } = this.props;

        return (
            <Form onSubmit={ handleSubmit }>
                <FormGroup
                    isRequired
                    label="Inventory data file"
                    fieldId="file"
                    helperTextInvalid={ errors.file }
                    isValid={
                        (errors.file && touched.file) ? false : true
                    }
                >
                    <Dropzone
                        onDrop={ (files: File[]) => {
                            this.props.onFileSelected(files);
                        } }
                        noClick noKeyboard
                        multiple={ false }
                        accept={ [ 'application/zip', 'application/json' ] }
                    >
                        { ({ getRootProps, getInputProps, open }) => {
                            return (
                                <div className="container">
                                    <div { ...getRootProps({ className: 'dropzone' }) }>
                                        <input { ...getInputProps() } />
                                        <InputGroup>
                                            <TextInput
                                                id="file"
                                                name="file"
                                                type="text"
                                                aria-label="Select a File"
                                                value={ values.file }
                                                isValid={
                                                    (errors.file && touched.file) ? false : true
                                                } />
                                            <Button
                                                variant={ ButtonVariant.secondary }
                                                aria-label="Browse a file to upload"
                                                onClick={ open }
                                            >
                                                Browse
                                            </Button>
                                        </InputGroup>
                                    </div>
                                </div>
                            );
                        } }
                    </Dropzone>
                </FormGroup>
                <FormGroup
                    isRequired
                    label="Report name"
                    fieldId="reportName"
                    helperTextInvalid={ errors.reportName }
                    isValid={
                        (errors.reportName && touched.reportName) ? false : true
                    }
                >
                    <TextInput
                        isRequired
                        type="text"
                        id="reportName"
                        name="reportName"
                        aria-describedby="Report name"
                        onChange={ (_value, event) => handleChange(event) }
                        onBlur={ handleBlur }
                        value={ values.reportName }
                        isValid={
                            (errors.reportName && touched.reportName) ? false : true
                        }
                    />
                </FormGroup>
                <FormGroup
                    label="Report description"
                    fieldId="reportDescription"
                    helperTextInvalid={ errors.reportDescription }
                    isValid={ (errors.reportDescription && touched.reportDescription) ? false : true }
                >
                    <TextInput
                        type="text"
                        id="reportDescription"
                        name="reportDescription"
                        aria-describedby="Report description"
                        onChange={ (_value, event) => handleChange(event) }
                        onBlur={ handleBlur }
                        value={ values.reportDescription }
                        isValid={
                            (errors.reportDescription && touched.reportDescription) ? false : true
                        }
                    />
                </FormGroup>
                <FormGroup
                    isRequired
                    label="Year-over-year growth rate for new hypervisors"
                    fieldId="yearOverYearGrowthRatePercentage"
                    helperTextInvalid={ errors.yearOverYearGrowthRatePercentage }
                    isValid={
                        (errors.yearOverYearGrowthRatePercentage && touched.yearOverYearGrowthRatePercentage) ? false : true
                    }
                >
                    <Gallery>
                        <GalleryItem>
                            <InputGroup>
                                <TextInput
                                    isRequired
                                    id="yearOverYearGrowthRatePercentage"
                                    type="number"
                                    name="yearOverYearGrowthRatePercentage"
                                    aria-describedby="Year-over-year growth rate for new hypervisors"
                                    className="pf-u-text-align-right"
                                    onChange={ (_value, event) => handleChange(event) }
                                    onBlur={ handleBlur }
                                    value={ values.yearOverYearGrowthRatePercentage }
                                    isValid={
                                        (errors.yearOverYearGrowthRatePercentage && touched.yearOverYearGrowthRatePercentage) ? false : true
                                    }
                                />
                                <InputGroupText className="percentage-box">%</InputGroupText>
                            </InputGroup>
                        </GalleryItem>
                    </Gallery>
                </FormGroup>
                <Stack gutter="sm" className="pf-sm-gutter">
                    <StackItem isFilled={ false }>
                        <FormGroup
                            label="Percentage of hypervisors migrated each year"
                            fieldId="percentageOfHypervisorsMigratedOnYear1"
                            helperTextInvalid={ errors.percentageOfHypervisorsMigratedOnYear1 || errors.percentageOfHypervisorsMigratedSum }
                            isValid={
                                (errors.percentageOfHypervisorsMigratedOnYear1 && touched.percentageOfHypervisorsMigratedOnYear1) ||
                                (errors.percentageOfHypervisorsMigratedSum) ? false : true
                            }
                        >
                            <Gallery>
                                <GalleryItem>
                                    <InputGroup>
                                        <InputGroupText className="year-box">Year 1</InputGroupText>
                                        <TextInput
                                            isRequired
                                            id="percentageOfHypervisorsMigratedOnYear1"
                                            type="number"
                                            name="percentageOfHypervisorsMigratedOnYear1"
                                            aria-label="Percentage of hypervisors migrated on year 1"
                                            className="pf-u-text-align-right"
                                            onChange={ (_value, event) => handleChange(event) }
                                            onBlur={ handleBlur }
                                            value={ values.percentageOfHypervisorsMigratedOnYear1 }
                                            isValid={
                                                (errors.percentageOfHypervisorsMigratedOnYear1
                                                    && touched.percentageOfHypervisorsMigratedOnYear1) ? false : true
                                            }
                                        />
                                        <InputGroupText className="percentage-box">%</InputGroupText>
                                    </InputGroup>
                                </GalleryItem>
                            </Gallery>
                        </FormGroup>
                    </StackItem>
                    <StackItem isFilled={ false }>
                        <FormGroup
                            label=""
                            fieldId=""
                            helperTextInvalid={ errors.percentageOfHypervisorsMigratedOnYear2 || errors.percentageOfHypervisorsMigratedSum }
                            isValid={
                                (errors.percentageOfHypervisorsMigratedOnYear2 && touched.percentageOfHypervisorsMigratedOnYear2) ||
                                (errors.percentageOfHypervisorsMigratedSum) ? false : true
                            }
                        >
                            <Gallery>
                                <GalleryItem>
                                    <InputGroup>
                                        <InputGroupText className="year-box">Year 2</InputGroupText>
                                        <TextInput
                                            isRequired
                                            id="percentageOfHypervisorsMigratedOnYear2"
                                            type="number"
                                            name="percentageOfHypervisorsMigratedOnYear2"
                                            aria-label="Percentage of hypervisors migrated on year 2"
                                            className="pf-u-text-align-right"
                                            onChange={ (_value, event) => handleChange(event) }
                                            onBlur={ handleBlur }
                                            value={ values.percentageOfHypervisorsMigratedOnYear2 }
                                            isValid={
                                                (errors.percentageOfHypervisorsMigratedOnYear2
                                                    && touched.percentageOfHypervisorsMigratedOnYear2) ? false : true
                                            }
                                        />
                                        <InputGroupText className="percentage-box">%</InputGroupText>
                                    </InputGroup>
                                </GalleryItem>
                            </Gallery>
                        </FormGroup>
                    </StackItem>
                    <StackItem isFilled={ false }>
                        <FormGroup
                            label=""
                            fieldId=""
                            helperTextInvalid={ errors.percentageOfHypervisorsMigratedOnYear3 || errors.percentageOfHypervisorsMigratedSum }
                            isValid={
                                (errors.percentageOfHypervisorsMigratedOnYear3 && touched.percentageOfHypervisorsMigratedOnYear3) ||
                                (errors.percentageOfHypervisorsMigratedSum) ? false : true
                            }
                        >
                            <Gallery>
                                <GalleryItem>
                                    <InputGroup>
                                        <InputGroupText className="year-box">Year 3</InputGroupText>
                                        <TextInput
                                            isRequired
                                            id="percentageOfHypervisorsMigratedOnYear3"
                                            type="number"
                                            name="percentageOfHypervisorsMigratedOnYear3"
                                            aria-label="Percentage of hypervisors migrated on year 3"
                                            className="pf-u-text-align-right"
                                            onChange={ (_value, event) => handleChange(event) }
                                            onBlur={ handleBlur }
                                            value={ values.percentageOfHypervisorsMigratedOnYear3 }
                                            isValid={
                                                (errors.percentageOfHypervisorsMigratedOnYear3
                                                    && touched.percentageOfHypervisorsMigratedOnYear3) ? false : true
                                            }
                                        />
                                        <InputGroupText className="percentage-box">%</InputGroupText>
                                    </InputGroup>
                                </GalleryItem>
                            </Gallery>
                        </FormGroup>
                    </StackItem>
                </Stack>

                <ActionGroup>
                    <Button variant="primary" type="submit" isDisabled={ isSubmitting } disabled={ isSubmitting }>Create report</Button>
                    <Button variant="secondary" type="button" onClick={ this.props.handleCancel }>Cancel</Button>
                </ActionGroup>
            </Form>
        );
    }
}

export default UploadForm;
