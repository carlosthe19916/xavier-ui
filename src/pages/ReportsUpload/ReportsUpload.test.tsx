import { shallow } from "enzyme";
import ReportsUpload from "./ReportsUpload";

const props = {
  file: null,
  success: null,
  error: null,
  progress: 100,
  uploading: true,

  uploadProgress: jest.fn(),
  uploadRequest: jest.fn(),
  selectUploadFile: jest.fn(),

  history: null,
  location: null,
  match: null
};

describe("ReportsUpload", () => {
  it("expect to render form", () => {
    const wrapper = shallow(<ReportsUpload { ...props }/>);
    expect(wrapper).toMatchSnapshot();
  });
});
