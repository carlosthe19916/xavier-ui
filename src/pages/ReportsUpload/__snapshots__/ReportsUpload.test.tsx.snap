// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportsUpload expect to render form 1`] = `
<Fragment>
  <PageHeader>
    <PageHeaderTitle
      title="Reports"
    />
  </PageHeader>
  <Connect(Main)>
    <Modal
      actions={Array []}
      ariaDescribedById="Report options"
      className=""
      hideTitle={false}
      isLarge={false}
      isOpen={true}
      isSmall={true}
      onClose={[Function]}
      showClose={true}
      title="Report options"
    >
      <Formik
        enableReinitialize={false}
        initialValues={
          Object {
            "file": undefined,
            "percentageOfHypervisorsMigratedOnYear1": 50,
            "percentageOfHypervisorsMigratedOnYear2": 30,
            "percentageOfHypervisorsMigratedOnYear3": 20,
            "percentageOfHypervisorsMigratedSum": 90,
            "reportDescription": "",
            "reportName": "",
            "yearOverYearGrowthRatePercentage": 5,
          }
        }
        isInitialValid={false}
        onSubmit={[Function]}
        validate={[Function]}
        validateOnBlur={true}
        validateOnChange={true}
      >
        <Component />
      </Formik>
    </Modal>
  </Connect(Main)>
</Fragment>
`;
